pgrx/src/nullable.rs

```diff
@@ -0,0 +1,520 @@
+//! Traits and implementations of functionality relating to Postgres null
+//! and nullable types, particularly in relation to arrays and other container
+//! types which implement nullable (empty slot) behavior in a
+//! "structure-of-arrays" manner
+
+use bitvec::slice::BitSlice;
+use std::{borrow::Cow, fmt::Debug, marker::PhantomData};
+
+pub enum Nullable<T> {
+    Valid(T),
+    Null,
+}
+
+impl<T> Debug for Nullable<T>
+where
+    T: Debug,
+{
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Valid(value) => f.debug_tuple("Valid").field(value).finish(),
+            Self::Null => write!(f, "Null"),
+        }
+    }
+}
+
+impl<T> Clone for Nullable<T>
+where
+    T: Clone,
+{
+    #[inline]
+    fn clone(&self) -> Self {
+        match self {
+            Self::Valid(arg0) => Self::Valid(arg0.clone()),
+            Self::Null => Self::Null,
+        }
+    }
+}
+
+impl<T> PartialEq for Nullable<T>
+where
+    T: Eq,
+{
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Self::Valid(l_value), Self::Valid(r_value)) => l_value == r_value,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+impl<T> Eq for Nullable<T> where T: PartialEq + Eq {}
+
+impl<T> PartialOrd for Nullable<T>
+where
+    T: PartialOrd + PartialEq + Eq,
```

Comment:
I don't think we need to have a stricter bound than necessary?

```suggestion
    T: PartialOrd
```

Reply:
PartialOrd requires PartialEq, but not Eq.
 I removed the Eq bound in 01e64fc3e01489ce402571f3504f80a96d2fe406